### ISSUE TITLE
DAG for running and storing results of long queries

### ISSUE THREAD

**Issue by sarayourfriend** [WordPress/openverse#4595]
## Problem

While Grafana (and other tools like it) provide good interfaces for visualising data, they are not good for long queries. Those tools rely on reasonable time outs configured for their data sources to prevent inexpert (or unaware) users of the tools from taking down a database or Elasticsearch cluster (for example) with a big query.

## Description

Instead, let's create a DAG for running these long queries. The DAG should start by being able to run queries against the catalog database, and it should be restricted to non-destructive queries. To that end, we can use a separate user to execute the query against the database (a user with readonly permissions).

The DAG should accept a string for the query and a note explaining the purpose of the query, and under what circumstances it can or cannot be terminated.

Basically, this would just be the "select_query" part of the batched update DAG.

## Alternatives

Maybe there are better ways for long running queries? I see some mention of using materialized views for them, but I can't find anything super clear on the matter.

@AetherUnbound @stacimc y'all's advice here would be helpful. The main thing I want to solve is not needing to open direct connections to the databases for long queries, with the understanding that some long queries are necessary and unavoidable.

**Comment by AetherUnbound** [comment]
This definitely seems reasonable, especially with all the other qualifiers/requirements you've mentioned. Given that we [have some interest in deploying Metabase](https://github.com/WordPress/openverse-infrastructure/issues/949), would this be supplanted by that in your mind? The advantage of having this in Airflow is that we can manage the timeouts a little better I imagine. We may also be able to trigger this DAG from other DAGs that need it and use the result in downstream DAGs, etc.

**Comment by sarayourfriend** [comment]
Metabase has similar problems with long queries, namely that you have leave your browser open (at least from what I understand) for the duration of the question execution: https://discourse.metabase.com/t/metabase-timeout-configuration/10302/4. That thread is about a reverse proxy connection timeout issue but you can see that closing the connection from the browser to Metabase also seems to prevent the query from completing. I can't really find a straight answer about it.

Metabase could be useful if for some reason Grafana isn't able to give us what we want, otherwise I think we can hold off deploying it so long as this problem of intentionally long queries is solved in a manageable way.

**Comment by AetherUnbound** [comment]
Ahh that's disappointing, thanks for sharing that!

**Comment by krysal** [comment]
> Metabase has similar problems with long queries

What a bummer! I was hoping that could be _the_ solution. Although having Grafana is great for short queries, I was fiddling with it the last few days and couldn't get the data for a query like this:

```sql
SELECT identifier FROM image WHERE license = 'pdm' AND license_version = '4.0';
```

I knew there were only 43 rows for it. It is understandable that Grafana cannot finish it since it has to scan the whole table and has a time limit. The DAG could help, but a tool that can export the result in formats like CSV or TSV would be ideal. We can also implement this on the DAG, of course. I just thought that if there are tools for it, then it is better to choose them than reinvent the wheel.

**Comment by sarayourfriend** [comment]
If the DAG produces a materialized view, that view can in turn be used in Grafana for visualisation. Support DAGs can be written to update such views as needed.

That being said, @krysal for that particular query you might be able to get the same information from Elasticsearch much faster than from the catalog DB.

**Comment by sarayourfriend** [comment]
Gotcha. That's a good feature requirement. The DAG could definitely upload a TSV to S3, either instead of or in addition to a materialised view for use in visualisations? It would depend on the purpose of the query, but a materialized view is a good intermediate step regardless, I believe, instead of e.g. streaming the output into memory in Airflow and uploading it, especially if that would just make the query dependent on the Airflow task capturing the output, making it non-resilient to redeployments or failures. The AWS S3 utilities may be useful for exporting the materialized view directly to S3?

**Comment by krysal** [comment]
@sarayourfriend One thing to note about materialized views is that they also take up space in the database, so queries that could double in size need to be monitored/cleaned. But yes, it is a good option, too.

> The AWS S3 utilities may be useful for exporting the materialized view directly to S3?

Yes, there is a [`aws_s3.query_export_to_s3`](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/postgresql-s3-export.html#aws_s3.export_query_to_s3) function to export to S3 😁

**Comment by sarayourfriend** [comment]
Good point too there, on the size of views. The DAG can have a "clean view after query" option if we just need the exported values.
